### PR TITLE
feat(exporter): in-node auditor to early detect missing duties and bugs

### PR DESCRIPTION
**Pull request by julienh-ssv**

### CODE REVIEW

**Comment by nkryuchkov** [comment]
@greptileai please review this again

File: api/handlers/exporter/auditor_http.go [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+type AuditorFindingsRequest struct {
+	From  *uint64 `json:"from,omitempty" minimum:"0"`
+	To    *uint64 `json:"to,omitempty" minimum:"0"`
+	LastN *uint64 `json:"lastN,omitempty" minimum:"0"`
+
+	Reason string `json:"reason,omitempty"`
+	Role   string `json:"role,omitempty"`
+
+	CommitteeID    string  `json:"committeeID,omitempty" format:"hex"`
+	ValidatorIndex *uint64 `json:"validatorIndex,omitempty" format:"int64"`
```
Comment (nkryuchkov) [inline api/handlers/exporter/auditor_http.go:26]: We have `uint64` type but `int64` format, can we add a test that negative numbers don't cause any issues? Or if `format` supports `uint64`, that would be better

File: api/handlers/exporter/auditor_http.go [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+	}
+
+	from, to, err := resolveSlotRange(req.From, req.To, req.LastN, lastAuditedSlot)
+	if err != nil {
+		return audit.Query{}, err
+	}
+
+	limit := req.Limit
+	if limit <= 0 {
+		limit = 100
```
Comment (nkryuchkov) [inline api/handlers/exporter/auditor_http.go:98]: Can we extract 100, 1000, `asc`, `desc` to consts?

File: api/handlers/exporter/auditor_http.go [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+	return api.Render(w, r, &AuditorStatusResponse{Data: st})
+}
+
+type AuditorSummaryRequest struct {
+	From  *uint64 `json:"from,omitempty" minimum:"0"`
+	To    *uint64 `json:"to,omitempty" minimum:"0"`
+	LastN *uint64 `json:"lastN,omitempty" minimum:"0"`
+
+	// MaxFindings limits how many stored findings are scanned to build the summary.
+	// This is a safety valve; defaults to 200000 and is capped to 500000.
```
Comment (nkryuchkov) [inline api/handlers/exporter/auditor_http.go:248]: This comment is easy to forget about if we decide to change those values, do you think we need it?

File: api/handlers/exporter/auditor_http.go [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+	}
+
+	from, to, err := resolveSlotRange(req.From, req.To, req.LastN, st.LastAuditedSlot)
+	if err != nil {
+		return toApiError(e.logger, r, "auditor_summary", http.StatusBadRequest, req, err)
+	}
+
+	maxFindings := req.MaxFindings
+	if maxFindings <= 0 {
+		maxFindings = 200000
```
Comment (nkryuchkov) [inline api/handlers/exporter/auditor_http.go:307]: Can we extract both 200000 and 500000 to consts?

File: operator/duties/attester.go [hunk 5]
```diff
@@ -44,6 +46,10 @@ func NewAttesterHandler(duties *dutystore.Duties[eth2apiv1.AttesterDuty], export
 	return h
 }
 
+func (h *AttesterHandler) SetAuditReporter(r audit.Reporter) {
```
Comment (nkryuchkov) [inline operator/duties/attester.go:49]: We will likely need to refactor how we set up the handler and pass the audit reporter as an optional parameter. But it should not be in the scope of this PR

File: operator/dutytracer/auditor/auditor.go [hunk 6]
```diff
@@ -0,0 +1,1384 @@
...
+	if epoch != nil {
+		return fmt.Sprintf("%s-e%d", role.String(), uint64(*epoch))
+	}
+	if period != nil {
+		return fmt.Sprintf("%s-p%d", role.String(), *period)
+	}
+	return role.String()
+}
+
+func (a *Auditor) AuditSlot(ctx context.Context, slot phase0.Slot) error {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/auditor.go:408]: `AuditSlot` is quite hard to read, can we decompose it?

File: operator/dutytracer/auditor/auditor.go [hunk 7]
```diff
@@ -0,0 +1,1384 @@
...
+}
+
+type reasonEvidence struct {
+	persistedSchedule ScheduleEvidence
+	registry          RegistryEvidence
+	links             LinksEvidence
+	pipeline          PipelineEvidence
+}
+
+func (a *Auditor) chooseReasonForScheduleMismatch(slot phase0.Slot, epoch phase0.Epoch, period uint64, role Role, index phase0.ValidatorIndex, persisted map[phase0.ValidatorIndex]rolemask.Mask, expectedByDutyStore bool, scheduleReadErr error, rpcUsed bool, rpcSkipped bool, rpcExpected bool, rpcExpectedSlot *uint64, rpcErr error, expectedOtherRole bool) (ReasonCode, reasonEvidence) {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/auditor.go:931]: Can we create a struct for all the stuff we pass, or refactor this anyhow?

File: operator/dutytracer/auditor/auditor.go [hunk 8]
```diff
@@ -0,0 +1,1384 @@
...
+		if expectedOtherRole {
+			return ReasonRoleClassificationSuspect, ev
+		}
+		return ReasonUnexpectedWireTrace, ev
+	}
+
+	return ReasonUnexpectedWireTrace, ev
+}
+
+func (a *Auditor) checkCommitteeMapping(slot phase0.Slot, epoch phase0.Epoch, period uint64, role Role, observedCommittee spectypes.CommitteeID, index phase0.ValidatorIndex, expectedCommittee map[phase0.ValidatorIndex]spectypes.CommitteeID, linksByIndex map[phase0.ValidatorIndex]spectypes.CommitteeID, persisted map[phase0.ValidatorIndex]rolemask.Mask, it obsItem, linksReadOK bool, linksReadErr error, emit func(*Finding) emitResult) error {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/auditor.go:1040]: ditto

File: operator/dutytracer/auditor/auditor.go [hunk 9]
```diff
@@ -0,0 +1,1384 @@
...
+	rpc := f.Evidence.Expected.RPCFallback
+	df := f.Evidence.Pipeline.DutyFetch
+	sc := f.Evidence.Pipeline.ScheduleCompute
+
+	var rpcOK any
+	if rpc.OK != nil {
+		rpcOK = *rpc.OK
+	}
+
+	fs := []zap.Field{
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/auditor.go:1219]: this is going to be quite a huge log, do we need all of these?

File: operator/dutytracer/auditor/store_db.go [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+	return append([]byte(indexCPrefixKey), raw...), nil
+}
+
+func makeReasonPrefix(reasonByte byte) []byte {
+	return append([]byte(indexRPrefixKey), reasonByte)
+}
+
+func makeIndexEntryKey(slot phase0.Slot, reasonByte byte, seq byte) []byte {
+	var b [6]byte
+	// Store slot in descending order using inverted uint32 big-endian.
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:123]: This comment repeats the code without explaining why. Can we rewrite it to explain why we are doing this rather than what we are doing?

File: operator/dutytracer/auditor/store_db.go [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+	seqU64, err := strconv.ParseUint(parts[2], 10, 64)
+	if err != nil {
+		return ci, fmt.Errorf("invalid cursor seq: %w", err)
+	}
+
+	rb := reasonToByte(ReasonCode(reasonStr))
+	if rb == 0 {
+		return ci, fmt.Errorf("invalid cursor reason: %s", reasonStr)
+	}
+	if seqU64 > 255 {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:168]: an optional nit:
```suggestion
	if seqU64 > math.MaxUint8 {
```

File: operator/dutytracer/auditor/store_db.go [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+func findingKey(slot phase0.Slot, reason ReasonCode, seq uint16) string {
+	return fmt.Sprintf("%d/%s/%d", uint64(slot), reason, seq)
+}
+
+func (s *DBStore) PutFinding(f *Finding) (PutResult, error) {
+	if f == nil {
+		return PutResult{}, fmt.Errorf("nil finding")
+	}
+	if f.CreatedAt.IsZero() {
+		f.CreatedAt = time.Now().UTC()
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:187]: We usually don't use `time.Now()` with `.UTC()`. Can you please make sure it won't cause any incompatibility issues with other usages of `time.Now()`?

File: operator/dutytracer/auditor/store_db.go [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+	}
+	if f.Version == 0 {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:189]: This might need a comment

File: operator/dutytracer/auditor/store_db.go [hunk 14]
```diff
@@ -0,0 +1,792 @@
...
+	if err != nil {
+		return PutResult{}, fmt.Errorf("marshal finding: %w", err)
+	}
+	if err := s.db.Set(prefixFinding, key, value); err != nil {
+		return PutResult{}, fmt.Errorf("save finding: %w", err)
+	}
+
+	// Increment the count.
+	var enc [2]byte
+	binary.LittleEndian.PutUint16(enc[:], count+1)
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:240]: Will it overflow if `count == 255`?

File: operator/dutytracer/auditor/store_db.go [hunk 15]
```diff
@@ -0,0 +1,792 @@
...
+				return QueryResult{}, fmt.Errorf("unsupported reason: %s", *q.Reason)
+			}
+			return s.queryByIndex(makeReasonPrefix(rb), q, cur)
+		}
+	}
+
+	return s.queryBySlotScan(q, cur)
+}
+
+func (s *DBStore) queryBySlotScan(q Query, cur cursorInfo) (QueryResult, error) {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:392]: can we decompose `queryBySlotScan`? it's quite complex to read

File: operator/dutytracer/auditor/store_db.go [hunk 16]
```diff
@@ -0,0 +1,792 @@
...
+	}
+	if q.ValidatorIndex != nil {
+		if f.ValidatorIndex == nil || *f.ValidatorIndex != *q.ValidatorIndex {
+			return false
+		}
+	}
+	return true
+}
+
+func minInt(a, b int) int {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:587]: We use a Go version with embedded `min`

File: operator/dutytracer/auditor/store_db.go [hunk 17]
```diff
@@ -0,0 +1,792 @@
...
+func makeSlotPrefix(base string, slot phase0.Slot) []byte {
+	var b [4]byte
+	// #nosec G115
+	binary.LittleEndian.PutUint32(b[:], uint32(uint64(slot)))
+	return append([]byte(base), b[:]...)
+}
+
+// reasonToByte provides a stable small key discriminator for a reason code.
+// Only reasons in the authoritative set are supported; unknown values return 0.
+func reasonToByte(r ReasonCode) byte {
```
Comment (nkryuchkov) [inline operator/dutytracer/auditor/store_db.go:744]: Do you think storing `ReasonCode` as `int` and mapping reason to text would be easier to read? E.g., `ReasonScheduleMissingIndex` -> `SCHEDULE_MISSING_INDEX`.

It's not clear why we are using exactly these numbers and why there are gaps

File: api/handlers/exporter/auditor_http.go [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+	}
+
+	// Default: last N slots ending at lastAuditedSlot.
+	if from == nil && to == nil && lastN == nil {
+		n := defaultLastN
+		if lastAuditedSlot < n {
+			return 0, lastAuditedSlot, nil
+		}
+		return lastAuditedSlot - n, lastAuditedSlot, nil
+	}
```
Comment (iurii-ssv) [inline api/handlers/exporter/auditor_http.go:176]: api/handlers/exporter/auditor_http.go:169 has an off-by-one bug in resolveSlotRange(). Both the default path and the lastN path compute start = end - n, but the range
     is inclusive on both ends, so lastN=256 returns 257 slots and lastN=1 returns 2 slots. The existing test already codifies the wrong behavior (744..1000 for a “last 256”
     request). That will over-read data and make pagination / API expectations inconsistent.

File: operator/dutytracer/auditor/auditor.go [hunk 19]
```diff
@@ -0,0 +1,1384 @@
...
+		return
+	}
+	defer func() {
+		if r := recover(); r != nil {
+			a.logger.Error("audit slot panicked", fields.Slot(slot), zap.Any("panic", r))
+		}
+	}()
+	if err := a.AuditSlot(ctx, slot); err != nil {
+		a.logger.Warn("audit slot failed", fields.Slot(slot), zap.Error(err))
+	}
```
Comment (iurii-ssv) [inline operator/dutytracer/auditor/auditor.go:272]: operator/dutytracer/auditor/auditor.go:270 updates lastAuditedSlot even when AuditSlot() failed or panicked. That means a transient failure on slot S is reported as
     successfully audited, the status endpoint advances past it, and the normal fixed-delay loop never retries that slot. Unless a late trace happens to arrive and trigger
     NotifyLateSlot, the gap is permanent and silent.

**Review by iurii-ssv — COMMENTED** [review]
Couple AI findings to check out, for now
